test/e2e/storage/testsuites/provisioning.go

```diff
@@ -0,0 +1,339 @@
+}
+
+func createProvisioningTestInput(pattern testpatterns.TestPattern, resource provisioningTestResource) provisioningTestInput {
+	driver := resource.driver
+	dInfo := driver.GetDriverInfo()
+	f := dInfo.Framework
+
+	input := provisioningTestInput{
+		testCase: StorageClassTest{
+			ClaimSize:    resource.claimSize,
+			ExpectedSize: resource.claimSize,
+		},
+		cs:  f.ClientSet,
+		pvc: resource.pvc,
+		sc:  resource.sc,
+	}
+
+	if dInfo.Config.ClientNodeName != "" {
+		input.testCase.NodeName = dInfo.Config.ClientNodeName
+	}
+
+	return input
+}
+
+func (p *provisioningTestSuite) execTest(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
+	Context(getTestNameStr(p, pattern), func() {
+		var (
+			resource     provisioningTestResource
+			input        provisioningTestInput
+			needsCleanup bool
+		)
+
+		BeforeEach(func() {
+			needsCleanup = false
+			// Skip unsupported tests to avoid unnecessary resource initialization
+			skipUnsupportedTest(p, driver, pattern)
+			needsCleanup = true
+
+			// Setup test resource for driver and testpattern
+			resource := provisioningTestResource{}
+			resource.setupResource(driver, pattern)
+
+			// Create test input
+			input = createProvisioningTestInput(pattern, resource)
+		})
+
+		AfterEach(func() {
+			if needsCleanup {
+				resource.cleanupResource(driver, pattern)
+			}
+		})
+
+		testProvisioning(&input)
+	})
+}
+
+type provisioningTestResource struct {
+	driver drivers.TestDriver
+
+	claimSize string
+	sc        *storage.StorageClass
+	pvc       *v1.PersistentVolumeClaim
+}
+
+var _ TestResource = &provisioningTestResource{}
+
+func (p *provisioningTestResource) setupResource(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
+	p.driver = driver
+
+	dInfo := p.driver.GetDriverInfo()
+	f := dInfo.Framework
+	ns := f.Namespace
+	volType := pattern.VolType
+
+	// Setup provisioningTest resource
+	switch volType {
+	case testpatterns.DynamicPV:
+		if dDriver, ok := driver.(drivers.DynamicPVTestDriver); ok {
+			p.sc = dDriver.GetDynamicProvisionStorageClass("")
+			if p.sc == nil {
+				framework.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", dInfo.Name)
+			}
+			p.claimSize = "2Gi"
+			p.pvc = getClaim(p.claimSize, ns.Name)
+			p.pvc.Spec.StorageClassName = &p.sc.Name
+			framework.Logf("In setup, pvc: %v", p.pvc)
+		}
+	default:
+		framework.Failf("Dynamic Provision test doesn't support: %s", volType)
+	}
+}
+
+func (p *provisioningTestResource) cleanupResource(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
+}
+
+type provisioningTestInput struct {
+	testCase StorageClassTest
+	cs       clientset.Interface
+	pvc      *v1.PersistentVolumeClaim
+	sc       *storage.StorageClass
+}
+
+func testProvisioning(input *provisioningTestInput) {
+	It("should provision storage", func() {
+		TestDynamicProvisioning(input.testCase, input.cs, input.pvc, input.sc)
```

Comment:
Is there another test suite this could be a part of? I'm not sure it's unique enough to warrant moving it out into its own suite.

Also if we're moving this to the new test suite framework can we make the in-tree drivers consume the tests suite in the new way as well?

Reply:
@davidz627

Thank you for your comment.

> Is there another test suite this could be a part of? I'm not sure it's unique enough to warrant moving it out into its own suite (but then again maybe it is, I didn't look that carefully).

```TestDynamicProvisioning``` is still used by below tests:
  - test/e2e/storage/regional_pd.go
  - test/e2e/storage/volume_provisioning.go

I expect that all of them will eventually move to this framework to ensure that all driver are tested in the same way.
However, I would like to limit the scope of this PR to make the testsuites of csi_volumes.go and in_tree_volumes.go the same. So, only provisioning tests that is called from csi_volumes.go was ported to this framework.
(There are so many cloud provider dependent tests in these files, so doing it at once will require too much work that is not the original purpose of this PR.)
 
Until then, ```TestDynamicProvisioning``` needs to be called from storage package. So, it is moved to testsuites package to be used in this framework and changed to global scope.


> Also if we're moving this to the new test suite framework can we make the in-tree drivers consume the tests suite in the new way as well?

It is already done in this commit and only adding below one line to test/e2e/storage/in_tree_volumes.go will make in-tree drivers run this provisioning test suite.

https://github.com/kubernetes/kubernetes/pull/68025/files/90dbd631243c70f7db3698410a67ea80b18cbf93#diff-e6441be49bc83f3d2436271f90c14742R53